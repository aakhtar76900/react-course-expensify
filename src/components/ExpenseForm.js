import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import {connect} from 'react-redux';
import { addExpense } from '../actions/expenses';


 class ExpenseForm extends React.Component {
    

    state = {
        description: this.props.expense.description ? this.props.expense.description : '',
        note: this.props.expense.note ? this.props.expense.note : '',
        amount: this.props.expense.amount ? this.props.expense.amount.toString() : '',
        createdAt: this.props.expense.amount? moment(this.props.expense.amount) : moment(),
        calendarFocused: false,
        error : undefined
    }

    handleDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }))
    }

    handleNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    handleAmountChange = (e) => {
        const amount = e.target.value;
        if (amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }

    handleOnDateChange = (createdAt) => {
        this.setState(() => ({ createdAt }));
    }

    handleOnFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }

    handleOnSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description)
        {
            this.setState(() => ({error: 'Please enter description'}));
            return false;
        }
        if(!this.state.amount)
        {
            this.setState(() => ({error: 'Please enter Amount'}));
            return false;
        }

        this.setState(() => ({error : undefined}));
        this.props.onSubmitForm({
            description : this.state.description ,
            amount : parseFloat(this.state.amount),
            createdAt : this.state.createdAt.valueOf(),
            note : this.state.note
        })

    }

    render() {
       
        return (
            <div>
                <form onSubmit={this.handleOnSubmit}>
                {this.state.error && <p>{this.state.error}</p>}
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.handleDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.handleAmountChange}
                    />
                    <textarea
                        placeholder="Enter Note"
                        value={this.state.note}
                        onChange={this.handleNoteChange}
                    ></textarea>

                    <div></div>

                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.handleOnDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={({focused}) => (this.handleOnFocusChange(focused))}
                        numberOfMonths={1}
                        isOutsideRange={(day)=> false}

                    />



                    <button>Add Expense</button>
                    
                </form>
            </div>
        )
    }
}

export default connect()(ExpenseForm);