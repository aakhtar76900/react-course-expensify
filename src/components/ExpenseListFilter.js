import React from 'react';
import { connect } from 'react-redux';
import {setFilterText, setSortByAmount, setSortByDate, setStartDate, setEndDate} from '../actions/filters';
import moment from 'moment';
import {DateRangePicker} from 'react-dates';


class ExpenseListFilter extends React.Component {

    state = {
        calendarFocused : null
    }

    handleOnDateChange = ({startDate, endDate}) => {
       
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    }

    handleFocusChange = (focusInput) => {
        this.setState(() => ({calendarFocused : focusInput}))
    }

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={(e)=> {
                    this.props.dispatch(setFilterText(e.target.value));
                }} />
                
                <select value={this.props.filters.sortBy} onChange={(e)=> {
                    if(e.target.value == 'amount')
                    {
                        this.props.dispatch(setSortByAmount());
                    }
                    else
                    {
                        this.props.dispatch(setSortByDate());
                    }
                }}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>

                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    startDateId="txtStartDateFilter"
                    endDate={this.props.filters.endDate}
                    endDateId="txtEndDateFilter"
                    onDatesChange={this.handleOnDateChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={ (focusedInput) => {this.handleFocusChange(focusedInput )} }
                    showClearDates={true}
                    isOutsideRange={() => false}
                    numberOfMonths={1}

                />

            </div>
        )
    }
}


const mapStateToProps = (state) => {
    return {
        filters : state.filters
    }
}



export default connect(mapStateToProps)(ExpenseListFilter);