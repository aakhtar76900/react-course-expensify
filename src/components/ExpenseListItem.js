import React from 'react';
import { removeExpense } from '../actions/expenses';
import { connect } from 'react-redux';
import {Link} from 'react-router-dom';


export const ExpenseListItem = (props) => {
    return (
        <div>
            <Link to={`/edit/${props.expense.id}`}>
            <h3>{props.expense.description} </h3></Link>
            <p>Amount : {props.expense.amount}</p>
            <p>Created At : {props.expense.createdAt}</p>
            <button onClick={() => {
                props.dispatch(removeExpense(props.expense.id));
            }} >Remove</button>
        </div>

    )
}


export default connect()(ExpenseListItem);