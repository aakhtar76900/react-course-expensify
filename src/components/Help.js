import React from 'react';

const Help = () => {
    return (
        <h3>Help Page</h3>
    )
}

export default Help;