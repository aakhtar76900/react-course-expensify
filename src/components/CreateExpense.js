import React from 'react';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';
import { connect } from 'react-redux';

const CreateExpense = (props) => {
    return (
        <div>
            <h3>Create Page</h3>
            <ExpenseForm expense={{}}
                onSubmitForm={(expense) => {
                    props.dispatch(addExpense(expense));
                    props.history.push('/');
                }} />
        </div>

    )
}

export default connect()(CreateExpense);