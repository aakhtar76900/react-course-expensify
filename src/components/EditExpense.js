import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense } from '../actions/expenses';

const EditExpense = ({match,expense,dispatch,history}) => {
    return (
        <div>
            <h3>Edit Page</h3>
            <p>You are editing item with id = {match.params.id} </p>
            <ExpenseForm 
                onSubmitForm={(updatedExpense) => {
                   dispatch(editExpense(expense.id,updatedExpense));
                    history.push('/');
                }}
                expense={expense}
            />
        </div>

    )
}

const mapStateToProps = (state,props) => {
    return {
        expense : state.expenses.find((expense) => (expense.id === props.match.params.id))
    };
    
}

export default connect(mapStateToProps)(EditExpense);