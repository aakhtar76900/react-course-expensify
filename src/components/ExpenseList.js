import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from '../components/ExpenseListItem';
import selectExpenses from '../selector/getVisibleExpense';

export const ExpenseList = (props) => {
    return (
        <div>
            {props.expenses.length === 0 ? (
                <p>No expenses available</p>
            ) : (props.expenses.map((expense) => {
                return (
                    <ExpenseListItem key={expense.id} expense={expense} />
                )
            }))}
        </div>
    )
}

const mapStateToProps = (state) => {
    return {
        name: 'Andrew jackson',
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList);

