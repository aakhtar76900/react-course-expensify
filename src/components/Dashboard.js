import React from 'react';
import {Link} from 'react-router-dom';
import ExpenseList from '../components/ExpenseList';
import ExpenseListFilter from '../components/ExpenseListFilter';

const Dashboard = () => {
    return (
        <div>
            <h3>Expenses Dashboard</h3>
            {/* <Link to="/edit/1"> Item 1 </Link> */}
            <ExpenseListFilter />
            <h4>List of Expenses</h4>
            <ExpenseList />
        </div>
    )
}

export default Dashboard;