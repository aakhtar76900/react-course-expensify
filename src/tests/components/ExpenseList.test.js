import React from 'react';
import {ExpenseList} from '../../components/ExpenseList';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';

test('should load expense list' , () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('should load expense list with empty array' , () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});