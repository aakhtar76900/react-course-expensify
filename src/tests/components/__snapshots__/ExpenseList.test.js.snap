// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should load expense list 1`] = `
<div>
  <Connect(ExpenseListItem)
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
    key="1"
  />
  <Connect(ExpenseListItem)
    expense={
      Object {
        "amount": 109500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    key="2"
  />
  <Connect(ExpenseListItem)
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
    key="3"
  />
</div>
`;

exports[`should load expense list with empty array 1`] = `
<div>
  <p>
    No expenses available
  </p>
</div>
`;
